File: strategy/src/main/java/com/nearsoft/workshop/strategy/character/Character.java
package com.nearsoft.workshop.strategy.character;

import com.nearsoft.workshop.strategy.weapon.WeaponBehaviour;

public abstract class Character {

    protected WeaponBehaviour weaponBehaviour;

    public abstract void fight();

    public void setWeaponBehaviour(WeaponBehaviour weaponBehaviour) {
        this.weaponBehaviour = weaponBehaviour;
    }
}

File: strategy/src/main/scala/com/nearsoft/workshop/strategy/scala/character/Character.scala
package com.nearsoft.workshop.strategy.scala.character

abstract class Character {
  var weaponBehaviour: Option[() => String] = None

  def fight

  def setWeaponBehaviour(weaponBehaviour: () => String) {
    this.weaponBehaviour = Some(weaponBehaviour)
  }

  def useWeapon = weaponBehaviour match {
    case Some(weapon) => weapon.apply()
    case None => "doing nothing, no weapon"
  }
}
